### PR TITLE
fix: poetry add --lock does not perform a full update

### PR DESCRIPTION
**Pull request by ralbertazzi**
# Pull Request Check List

Resolves: #7309 

~It builds on top of https://github.com/python-poetry/poetry/pull/7917 . Once merged I'll rebase accordingly.~ ✅ 

- [x] Added **tests** for changed code.
- ~[ ] Updated **documentation** for changed code.~

### CODE REVIEW

File: tests/fixtures/up_to_date_lock/pyproject.toml [hunk 1]
```diff
@@ -6,7 +6,7 @@ authors = ["Poetry Developer <developer@python-poetry.org>"]
 
 [tool.poetry.dependencies]
 python = "^3.8"
-docker = "4.3.1"
+docker = ">=4.3.1"
```
Comment (ralbertazzi) [inline tests/fixtures/up_to_date_lock/pyproject.toml:9]: This was necessary for `test_add_does_not_update_locked_dependencies` to let the solver try to (wrongly) update from `4.3.1` to `4.3.2` when `--lock` is specified. I verified that the test was failing with the previous implementation and passes now

File: src/poetry/console/commands/add.py [hunk 2]
```diff
@@ -266,8 +264,7 @@ def handle(self) -> int:
         self.installer.dry_run(self.option("dry-run"))
         self.installer.verbose(self.io.is_verbose())
         self.installer.update(True)
-        if self.option("lock"):
-            self.installer.lock()
```
Comment (ralbertazzi) [inline src/poetry/console/commands/add.py:270]: In another PR I might remove `Installer.lock()` entirely as it's quite confusing. Its behaviour is in fact not in sync with the `--lock` options of the Poetry commands

Comment (radoering) [reply src/poetry/console/commands/add.py:270]: Yeah, quite confusing. IIRC, it's only set for `poetry lock`.

File: tests/console/commands/test_add.py [hunk 3]
```diff
@@ -1506,3 +1511,29 @@ def test_add_extras_only_accepts_one_package(
...
+def test_add_does_not_update_locked_dependencies(
+    repo: TestRepository,
+    poetry_with_up_to_date_lockfile: Poetry,
+    tester: CommandTester,
+    command_tester_factory: CommandTesterFactory,
+    command: str,
+) -> None:
+    tester = command_tester_factory("add", poetry=poetry_with_up_to_date_lockfile)
+    docker_locked = get_package("docker", "4.3.1")
+    dockr_new = get_package("docker", "4.3.2")
```
Comment (radoering) [inline tests/console/commands/test_add.py:1526]:
```suggestion
    docker_new = get_package("docker", "4.3.2")
```

File: src/poetry/console/commands/add.py (hunk 2 above)
Comment (ralbertazzi) [reply src/poetry/console/commands/add.py:270]: It is now ;)

File: tests/console/commands/test_add.py [hunk 4]
```diff
@@ -1506,3 +1511,29 @@ def test_add_extras_only_accepts_one_package(
             str(e.value)
             == "You can only specify one package when using the --extras option"
         )
+
+
+@pytest.mark.parametrize("command", ["foo", "foo --lock"])
```
Comment (radoering) [inline tests/console/commands/test_add.py:1516]:
```suggestion
@pytest.mark.parametrize("command", ["foo", "foo --lock"])
@pytest.mark.parametrize(("locked", "expected_docker"), [(True, "4.3.1"), (False, "4.3.2")])
```

Thinking about the `poetry.locker.locked(True)`, I just came up with an idea to make the test more robust.

Comment (ralbertazzi) [reply tests/console/commands/test_add.py:1512]: Great suggestion! I just added it ;)

**Review by radoering — APPROVED** [review]

**Pull request merged**